fix(react): updated textarea counter value changes on re-render

TannerS opened the pull request:
Closes #

https://github.com/carbon-design-system/carbon/issues/12348

#### Changelog

**New**

- Added more unit test

**Changed**

- Added useEffect for value when changed, will update counter

#### Testing / Reviewing
I actually had to use the sandbox code and manually insert it into the storybook for local testing but its not something that will be testable in storybook 

NOTE: will conflict with test file here: https://github.com/carbon-design-system/carbon/pull/12906/files#diff-67daf23818c005abac39307f47f299b05283736cfede71b034c3b2d127f48fcf

Commit:
fix(react): updated textarea counter value changes on re-render

francinelucca commented on packages/react/src/components/TextArea/__tests__/TextArea-test.js:
```diff
@@ -10,30 +10,302 @@ import TextArea from '../TextArea';
+      expect(screen.getByRole('textbox')).toHaveAttribute('rows', '25');
+    });
+
+    it('should respect enableCounter and macCount prop', () => {
```
```suggestion
    it('should respect enableCounter and maxCount prop', () => {
```

francinelucca reviewed (changes requested):
Added this story into the `TextArea` story file:
```
export const Test = () => {
  return (<>
  <TextInput
    labelText="TextInputLabel"
    placeholder="Type something here"
    onChange={(e) =>
      (document.getElementById("text-area").value = e.target.value)
    }
  />
  <br></br>
  <br></br>
  <TextArea
    id="text-area"
    labelText="TextAreaLabel"
    maxCount={500}
    enableCounter
    placeholder="Counter should be updated when value is changed by typing something in the TextInput"
  />
</>)
}
```
 and loaded my local storybook and did not see the value update upon typing:
<img width="1539" alt="image" src="https://user-images.githubusercontent.com/40550942/229560515-c806cd53-625a-48ef-87e1-0b8ef4abfa28.png">

francinelucca committed:
fix: format

andreancardona commented on packages/react/src/components/TextArea/__tests__/TextArea-test.js:
```diff
@@ -10,30 +10,302 @@ import TextArea from '../TextArea';
+
+    it('should respect id prop', () => {
+      render(<TextArea id="textarea-1" labelText="TextArea label" />);
+
```
```suggestion
```

andreancardona commented on packages/react/src/components/TextArea/__tests__/TextArea-test.js:
```diff
@@ -10,30 +10,302 @@ import TextArea from '../TextArea';
+
+    it('should respect rows prop', () => {
+      render(<TextArea id="textarea-1" labelText="TextArea label" rows={25} />);
+
```
```suggestion
```

andreancardona commented on packages/react/src/components/TextArea/__tests__/TextArea-test.js:
```diff
@@ -10,30 +10,302 @@ import TextArea from '../TextArea';
+          maxCount={500}
+        />
+      );
+
```
```suggestion
```

andreancardona committed:
Update packages/react/src/components/TextArea/__tests__/TextArea-test.js

andreancardona committed:
Update packages/react/src/components/TextArea/__tests__/TextArea-test.js

andreancardona committed:
Update packages/react/src/components/TextArea/__tests__/TextArea-test.js

TannerS committed:
Update packages/react/src/components/TextArea/__tests__/TextArea-test.js

Co-authored-by: Francine Lucca <40550942+francinelucca@users.noreply.github.com>

Commit:
fix(react): fixed textarea test with updated rtl

Commit:
Merge remote-tracking branch 'origin/tanner_fixTextAreaCounter' into tanner_fixTextAreaCounter

francinelucca reviewed (changes requested):
See https://github.com/carbon-design-system/carbon/pull/13449#issuecomment-1511868600

TannerS committed:
Merge branch 'main' into tanner_fixTextAreaCounter

andreancardona committed:
Merge branch 'main' into tanner_fixTextAreaCounter

francinelucca committed:
fix(TextArea): use textarea ref value instead of [value,defaultValue]

francinelucca committed:
fix: format

francinelucca committed:
Merge branch 'main' of github.com:carbon-design-system/carbon into tanner_fixTextAreaCounter

francinelucca committed:
fix(TextArea): add value to textCounter dependency array

francinelucca committed:
fix(TextArea): textCount optimizations

- use value,defaultValue in useEffect insead of ref
- delay textCount assignation

francinelucca committed:
fix: format

andreancardona reviewed (approved):
reviewed and looks good!

andreancardona committed:
Merge branch 'main' into tanner_fixTextAreaCounter

kodiakhq[bot] committed:
Merge branch 'main' into tanner_fixTextAreaCounter
